core/src/main/java/org/testcontainers/utility/DockerImageName.java

```diff
@@ -159,63 +176,86 @@ public String getRegistry() {
     }
 
-    private interface Versioning {
-        boolean isValid();
-
-        String getSeparator();
+    /**
+     * Declare that this {@link DockerImageName} is a compatible substitute for another image - i.e. that this image
+     * behaves as the other does, and is compatible with Testcontainers' assumptions about the other image.
+     *
+     * @param otherImageName the image name of the other image
+     * @return an immutable copy of this {@link DockerImageName} with the compatibility declaration attached.
+     */
+    public DockerImageName asCompatibleSubstituteFor(String otherImageName) {
+        return asCompatibleSubstituteFor(DockerImageName.parse(otherImageName));
     }
 
-    @Data
-    private static class TagVersioning implements Versioning {
-        public static final String TAG_REGEX = "[\\w][\\w\\.\\-]{0,127}";
-        private final String tag;
-
-        TagVersioning(String tag) {
-            this.tag = tag;
-        }
-
-        @Override
-        public boolean isValid() {
-            return tag.matches(TAG_REGEX);
-        }
-
-        @Override
-        public String getSeparator() {
-            return ":";
-        }
-
-        @Override
-        public String toString() {
-            return tag;
-        }
+    /**
+     * Declare that this {@link DockerImageName} is a compatible substitute for another image - i.e. that this image
+     * behaves as the other does, and is compatible with Testcontainers' assumptions about the other image.
+     *
+     * @param otherImageName the image name of the other image
+     * @return an immutable copy of this {@link DockerImageName} with the compatibility declaration attached.
+     */
+    public DockerImageName asCompatibleSubstituteFor(DockerImageName otherImageName) {
+        return new DockerImageName(rawName, registry, repo, versioning, otherImageName);
     }
 
-    @Data
-    private static class Sha256Versioning implements Versioning {
-        public static final String HASH_REGEX = "[0-9a-fA-F]{32,}";
-        private final String hash;
-
-        Sha256Versioning(String hash) {
-            this.hash = hash;
+    /**
+     * Test whether this {@link DockerImageName} has declared compatibility with another image (set using
+     * {@link DockerImageName#asCompatibleSubstituteFor(String)} or
+     * {@link DockerImageName#asCompatibleSubstituteFor(DockerImageName)}.
+     * <p>
+     * If a version tag part is present in the <code>other</code> image name, the tags must exactly match, unless it
+     * is 'latest'. If a version part is not present in the <code>other</code> image name, the tag contents are ignored.
+     *
+     * @param other the other image that we are trying to test compatibility with
+     * @return whether this image has declared compatibility.
+     */
+    public boolean isCompatibleWith(DockerImageName other) {
```

Comment:
WDYT about adding `if (equals(other)) return true`?

Reply:
It doesn't decrease the amount of code, as we still have to do some form of wildcarding on the tag, so I think I'm inclined to leave as-is right now.

Reply:
but if it is the same instance, we don't need to check anything, do we?

Reply:
Sorry, what I meant was that we can't get rid of these lines:
```
		final boolean thisRegistrySame = other.registry.equals(this.registry);
        final boolean thisRepoSame = other.repo.equals(this.repo);
        final boolean thisVersioningNotSpecifiedOrSame = other.versioning == null ||
            other.versioning.equals(Versioning.TagVersioning.LATEST) ||
            other.versioning.equals(this.versioning);

        if (thisRegistrySame && thisRepoSame && thisVersioningNotSpecifiedOrSame) {
            return true;
        }
```
because the version comparison is looser than what `equals` does.

So if we add an `equals` or `==` check first then yes, we could return from the method more quickly when there's a narrower match.

But we're not optimising for performance, so I'm not sure that we need to. i.e. if there's equality, the existing code still does the right thing.

Reply:
Ok, we can keep it as is 👍